rules/Transform/Rector/Class_/AddAllowDynamicPropertiesAttributeRector.php

```diff
@@ -0,0 +1,134 @@
+use Rector\FamilyTree\Reflection\FamilyRelationsAnalyzer;
+use Rector\Php80\NodeAnalyzer\PhpAttributeAnalyzer;
+use Rector\PhpAttribute\Printer\PhpAttributeGroupFactory;
+use Rector\VersionBonding\Contract\MinPhpVersionInterface;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://wiki.php.net/rfc/deprecate_dynamic_properties
+ *
+ * @see \Rector\Tests\Transform\Rector\Class_\AddAllowDynamicPropertiesAttributeRector\AddAllowDynamicPropertiesAttributeRectorTest
+ */
+final class AddAllowDynamicPropertiesAttributeRector extends AbstractRector implements MinPhpVersionInterface
+{
+    /**
+     * @var string
+     */
+    private const ATTRIBUTE = 'AllowDynamicProperties';
+
+    public function __construct(
+        private FamilyRelationsAnalyzer $familyRelationsAnalyzer,
+        private PhpAttributeAnalyzer $phpAttributeAnalyzer,
+        private PhpAttributeGroupFactory $phpAttributeGroupFactory,
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add the `AllowDynamicProperties` attribute to all classes', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+class SomeObject {
+    public string $someProperty = 'hello world';
+}
+CODE_SAMPLE
+
+                ,
+                <<<'CODE_SAMPLE'
+#[AllowDynamicProperties]
+class SomeObject {
+    public string $someProperty = 'hello world';
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (
+            $this->isDescendantOfStdclass($node) ||
+            $this->hasNeededAttributeAlready($node) ||
+            $this->hasMagicSetMethod($node)
+        ) {
+            return null;
+        }
+
+        return $this->addAllowDynamicPropertiesAttribute($node);
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::DEPRECATE_DYNAMIC_PROPERTIES;
+    }
+
+    private function isDescendantOfStdclass(Class_ $node): bool
+    {
+        if (!$node->extends instanceof FullyQualified) {
+            return false;
+        }
+
+        $ancestorClassNames = $this->familyRelationsAnalyzer->getClassLikeAncestorNames($node);
+        return in_array('stdClass', $ancestorClassNames);
+    }
+
+    private function hasNeededAttributeAlready(Class_ $node): bool
+    {
+        $nodeHasAttribute = $this->phpAttributeAnalyzer->hasPhpAttribute($node, self::ATTRIBUTE);
+        if ($nodeHasAttribute === true) {
+            return true;
+        }
+
+        if (!$node->extends instanceof FullyQualified) {
+            return false;
+        }
+
+        // TODO: recursive check for if ancestors applied the attribute
+        $parentNode = $this->getParentClassNode($node);
```

Comment:
@samsonasik - This is where I made an intentionally incorrect change to explain what I'm imagining.

Essentially I want to be able to retrieve the parent class to the current node, and have it returned as a node. Then we can pass that back to the method recursively until we hit either: a) the attribute we're looking for, or b) the root class.

Reply:
For this concern -checking attributes inherited from the ancestors- I'm wondering if a similar technique of using reflection will work? I know that the native reflection class has a `getAttributes` method for this purpose.

However I'm still digging into what we have access to here - I don't see an equivalent method on the `ClassReflection` provided by PHPstan yet.

Reply:
If there's not an existing analyzer better suited, I'm wondering if it makes sense to add a method like:
`PhpAttributeAnalyzer::hasInheritedPhpAttribute(ClassLike $node, string $attributeClass): bool`

That can determine if the current `ClassLike` node's ancestors apply the attribute class we're looking for.

Reply:
You may need to collect all ancestors first, then loop to get class and grab via `AstResolver::resolveClassFromName`

https://github.com/rectorphp/rector-src/blob/9ba0f6305f9f1bce54ba73afbacad2d62ccc1022/src/PhpParser/AstResolver.php#L73


```php
$ancestorClassReflections = $classReflection->getAncestors();

        foreach ($ancestorClassReflections as $ancestorClassReflection) {
            $ancestorClassName = $ancestorClassReflection->getName();
            if ($ancestorClassName === $className) {
                continue;
            }
           
            $class = $this->astResolver->resolveClassFromName($ancestorClassName);
            if (! $class instanceof Class_) {
                continue;
           } 

            // check attribute exists heree
          ....
        }
```

Reply:
@samsonasik - thank you so much for that advice and all the coaching while creating this rule. i've added the new method as mentioned and it seems to work great.